docs/development_guidelines.md

```diff
@@ -76,3 +76,38 @@ but also helps fellow contributors making adjustments to your code in the future
 By default, Home Assistant will statically check for type hints in our automated CI process.
 Python modules can be included for strict checking, if they are fully typed, by adding an entry
 to the `.strict-typing` file in the root of the Home Assistant Core project.
+
+### Assertions
+
+Use `assert` statements only for type annotations to help mypy, and for pytest assertions. This is because assertions can be turned off so they cannot be relied upon for runtime checks. For error handling use `if` and log then `return` or `raise` an exception as appropriate.
+
+Good, an assertion to help mypy:
+```python
+def foo(bar: Bar) -> None:
+    assert bar.baz is not None
```

Comment:
There are cases where type checking is not able to track it's past checks and assert helps guiding typing.

```python
optional_var: Something | None

if self.optional_var is None:
    self._do_thing_without_optional_var()
else:
    self._do_thing_with_optional_var()
```

In above example `self._do_thing_with_optional_var()` would need an assert because typing thinks `optional_var` to be None when it's called.

Using `assert` that is not supposed to be hit in this case is preferred over using an if-statement because in the case of an if-statement, you need to handle/test the case where it's false.

In TypeScript they allow to use `!` at the end of a variable to indicate you know it exists.

Reply:
Thank you for reading over this. As balloob said, unfortunately mypy isn't smart enough (or doesn't have the entire context) to know everything about a variable's type, and sometimes has to be helped with assertions.

I've update the example for type checking assertions, based on balloob's example and with a bit more surrounding context.